recipe/meta.yaml

```
@@ -58,6 +58,8 @@ requirements:
     - cuda-cudart-dev
     - cuda-driver-dev
 {% endif %}
+  run:
+    - __cuda  # [cuda_compiler_version != "None"]
```

There might be smarter Jinja/selector tricks to achieve this, but this is one way.
```suggestion
  run_constrained:
    - __cuda <0.a0     # [cuda_compiler_version == "None"]
    - __cuda >=11.0.*  # [(cuda_compiler_version or "").startswith("11")]
    - __cuda >=12.0.*  # [(cuda_compiler_version or "").startswith("12")]
```

---

I indeed had [issues](https://github.com/conda-forge/pennylane-feedstock/pull/40) with CF CIs and the new Kokkos package, but I don't see why would the `run` condition above break CIs?

---

`run: __cuda` means that the CUDA driver must be present, or at least be hacked to present (via `CONDA_OVERRIDE_CUDA` that is set in the CUDA builds). For CPU only packages, `__cuda` is not available (or is set to 0, I forgot the behavior), and this dependency would not be satisfied.

---

But in CPU only packages, wouldn't `# [cuda_compiler_version != "None"]` evaluate to `False` and then `run: __cuda` should not be present?

---

That's why I made the suggestion 🙂 The current PR would break downstream

---

Sorry, I meant if `# [cuda_compiler_version != "None"]` evaluates to `False`, then it is as if `run: __cuda` is not in the file, and hence there should be no problem.

---

Sorry, it's been a long day... I meant to say the current PR would not offer enough constraints for the solver, when a machine has the CUDA driver installed. It may still be possible to install the CPU only package. 

In the past, packages like UCX used to have a way to prioritize over a certain flavor. (An analogy is the "nompi" builds such as HDF5.) Those tricks are gradually being removed in recent years, because many of the build systems are made such that a coexisting build is possible. A notable example is Open MPI, where it's by design safe to build all support (CPU/GPU) and have the latter opt-in at run time. Kokkos is clearly not the case, so I remain concerned.

---

Thanks a bunch @leofang, I wanted to make sure I had the right understanding of the issue. It appears to be a solver issue first and foremost. As for the solution, I think what you suggested would force the solver into doing what we want. Cheers.